programs/marginfi/src/state/emode.rs

```diff
@@ -0,0 +1,440 @@
+    pub asset_weight_maint: WrappedI80F48,
+}
+
+impl EmodeEntry {
+    pub fn is_empty(&self) -> bool {
+        self.collateral_bank_emode_tag == EMODE_TAG_EMPTY
+    }
+    pub fn tag_equals(&self, tag: u16) -> bool {
+        self.collateral_bank_emode_tag == tag
+    }
+}
+
+/// Users who borrow multiple e-mode assets at the same time get the LEAST FAVORABLE treatment
+/// between the borrowed assets, regardless of the amount of each asset borrowed. For example, if
+/// borrowing an LST and USDC against SOL, the user would normally get an emode benefit for LST/SOL,
+/// but since they are also borrowing USDC, they get only standard rates.
+///
+/// Returns the INTERSECTION of EmodeConfigs. If passed configs have the same
+/// `collateral_bank_emode_tag`, we will take the one that has:
+/// 1) The lesser of the flags (e.g. if just one has APPLIES_TO_ISOLATED, we take flags = 0)
+/// 2) The lesser of both asset_weight_init and asset_weight_maint
+///
+/// If one config has a collateral_bank_emode_tag and the others do not, ***we don't make an
+/// EmodeEntry for it at all***, i.e. there is no benefit for that collateral
+///
+/// ***Example 1***
+/// * entry | tag | flags | init | maint
+/// * 0       101    1       70     75
+/// * 1       101    0       60     80
+/// Result
+/// * tag | flags | init | maint
+/// * 101    0       60     75
+///
+///
+/// ***Example 2***
+/// * entry | tag | flags | init | maint
+/// * 0       99     1       70     75
+/// * 1       101    0       60     80
+/// Result
+/// * tag | flags | init | maint
+/// * empty
+///
+///
+/// ***Example 3***
+/// * entry | tag | flags | init | maint
+/// * 0       101    1       70     75
+/// * 1       101    0       60     80
+/// * 2       101    0       60     80
+/// * 2       99     0       60     80
+/// Result
+/// * tag | flags | init | maint
+/// * 101    0       60     75
+pub fn reconcile_emode_configs(configs: Vec<EmodeConfig>) -> EmodeConfig {
+    // TODO benchmark this in the mock program
+    // If no configs, return a zeroed config.
+    if configs.is_empty() {
+        return EmodeConfig::zeroed();
+    }
+    // If only one config, return it.
+    if configs.len() == 1 {
+        return configs.into_iter().next().unwrap();
+    }
+
+    let num_configs = configs.len();
+    // Stores (tag, (entry, tag_count)), where tag_count is how many times we've seen this tag. This
+    // BTreeMap is logically easier on the eyes, but is probably fairly CU expensive, and should be
+    // benchmarked at some point, a simple Vec might actually be more performant here
+    let mut merged_entries: std::collections::BTreeMap<u16, (EmodeEntry, usize)> =
+        std::collections::BTreeMap::new();
+
+    for config in &configs {
+        for entry in config.entries.iter() {
+            if entry.is_empty() {
+                continue;
+            }
+            // Note: We assume that entries is de-duped and each tag appears at most one time!
```

Comment:
- [ ] I'm assuming this is actually enforced when adding new entries but just reminding myself to check

Reply:
yes, this happens in `validate_entries`